File: Python/validPalindrome.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: validPalindrome.py
@Function: 验证回文字符串 Ⅱ 双指针 + 贪心策略
@Link: https://leetcode-cn.com/problems/valid-palindrome-ii/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-08
"""

class Solution:
    def validPalindrome(self, s):
        def checkPalindrome(low, high):
            i, j = low, high
            while i < j:
                if s[i] != s[j]:
                    return False
                i += 1
                j -= 1
            return True

        low, high = 0, len(s) - 1
        while low < high:
            if s[low] == s[high]: 
                low += 1
                high -= 1
            else:
                return checkPalindrome(low + 1, high) or checkPalindrome(low, high - 1)
        return True


# -------------------------
if __name__ == "__main__":
    # string = "aba"
    # string = "abca"
    string = "abcfdba"
    

    solution = Solution()
    check = solution.validPalindrome(string)
    print(f"The solution of this problem is {check}")

File: C++/validPalindrome.cpp
/**
 * @File    : validPalindrome.cpp
 * @Brief   : 验证回文字符串 Ⅱ 双指针 + 贪心策略
 * @Link    : https://leetcode-cn.com/problems/valid-palindrome-ii/
 * @Author  : Wei Li
 * @Date    : 2021-07-08
*/

#include <iostream>
#include <string>

class Solution
{
public:
    bool checkPalindrome(const std::string& s, int low, int high)
    {
        for (int i = low, j = high; i < j; ++i, --j)
        {
            if (s[i] != s[j])
            {
                return false;
            }
        }
        return true;
    }

    bool validPalindrome(std::string s)
    {
        int low = 0, high = s.size() - 1;
        while (low < high)
        {
            char c1 = s[low], c2 = s[high];
            if (c1 == c2)
            {
                ++low;
                --high;
            }
            else
            {
                return checkPalindrome(s, low, high - 1) || checkPalindrome(s, low + 1, high);
            }
        }
        return true;
    }
};

// -------------------------------
int main(int argc, char **argv)
{
    // std::string str = "abcfdba";
    std::string str = "aba";
    // std::string str = "abca";

    auto solution = Solution();
    bool check = solution.validPalindrome(str);
    std::cout << "The solution of this problem is : " << std::boolalpha << check << std::endl;
    
    return 0;
}
